### PR TITLE
[Reviewed + Merged] backfilling remaining data and test suite

### PR DESCRIPTION
**Pull request by cjm771**
- [x]  serve remaining routes: POST {trailId}/paths 
- [x] do sorting for paths: /GET {trailId}/recordings?sortBy={date|ranking},{asc|desc}
- [x] fill 21 + 100 with backfill data since these entries dont exist
- [x] for paths that we dont have on S3, backfill this data from a set of data that we do have
- [x] supply sufficient coverage for unit tests
- [x] supply sufficient coverage for integretion tests

### CODE REVIEW

**Comment by MengRS** [comment]
Beginning review now.

File: services/validator.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    {
+      name: 'tag',
+      type: 'string',
+      afterValidate: (val) => {
+        return val.trim();
+      }
+    },
+    {
+      name: 'ranking',
```
Comment (MengRS) [inline services/validator.js]: @cjm771, curious why this is ranking and not rating for consistency sake?

**Review by MengRS — CHANGES_REQUESTED** [review]
@cjm771, completed review - see notes below... please review the unchecked boxes.  BTW - great job on error handling all around - way more than I did.

# General Notes on API:

- [ ] Seems like certain .gpx links do not work - I'm assuming that this is expected behavior as you need to be logged in / authenticated to see those files?


> - Chris: yes this is the case for any entries that say `have_gpx: false`. These get backfilled on their respective `api_path_url` endpoint with a random seeded gpx file that we do have. (See: https://github.com/rpt09-scully/paths-service/tree/01_backfill_data_and_test_suite#153-backfilling-entries-via-a-seededrandom) I suppose i can also supply the backfilled_url for the fake gpx path, but likely i will not need this and its not public facing! But good catch. 

- [ ] May want to consider updating and putting the data shape of each particular path in your example-data.json (including the POST api endpoint - expected body parameters)


> - Chris : GOod thought Ill do this.


# API Testing
- [x] /paths 
`Comments:` Works great

- [ ] /:trailId/paths?sortBy={id|rating|date},{asc|desc}* 
`Comments`: Not sure if the `sort` works - can you double check to see if I entered the url correct? 
image

- [ ] /:trailId/recordings?sortBy={id|rating|date},{asc|desc}*
`Comments:` Same note on above on date sorting...

> - Chris: interesting, ill look into this, it seemed to be passing the tests.

- [ ] (POST) /:trailId/recordings
`Comments:` Can't get this working in postman - tried `form-data`, `x-www-form-urlencoded` and `raw`.  What am I doing wrong?


 > - Chris:  the server is expecting json and also the `xml` property is looking for an actual xml syntax string, not a url. The intention would be once, xml is validated it would be uploaded to S3. See last test in `test/integration/endpoints.test.js` for valid xml

- [x] /:trailId/heroPath
`Comments:` Works great

- [x] /paths/:pathId
`Comments:` Works great

- [x] /:trailId/trailHead
`Comments:` Works great

- [x] 100+ entries in database

File: services/validator.js (hunk 1 above)
Comment (cjm771) [reply services/validator.js]: should be rating ill revise

**Comment by cjm771** [comment]
Wow @MengRS thanks for the in depth review gifs and all! I edited your review with my comments and will revise, Not sure what's going on with sorting but ill investigate. 

Also did you manage to run and pass all the jest tests? Thanks

**Comment by MengRS** [comment]
@cjm771 - thank you for the comments / clarification above!  I tested your post api endpoint now (thanks for the tip!) I must be doing something wrong, can you see why [this](https://imgur.com/a/uhX7f53) does not post into the database?

And yes, all your jest tests passed on my machine. 💃

**Comment by cjm771** [comment]
@MengRS looks like its accepting the post (i.e. no errors), i havent implemented the actual saving of the content to db as of yet. see TODO here:  https://github.com/rpt09-scully/paths-service/tree/01_backfill_data_and_test_suite#12-to-do

 See revisions i just made, realized that for dates i needed to do some preformatting before sorting so added extra tests and stuff.

**Comment by MengRS** [comment]
@cjm771, ah! thanks for the note on the post! I need to learn how to read better.  

tested your date sort - works perfectly.

**Comment by cjm771** [comment]
Thanks ! Yeah no it was hidden in there, props to you for being so thorough✌🏽

Kind Regards,
Chris Malcolm

**Pull request merged**